### ISSUE TITLE
Person identity: face embedding + body re-ID for person-vs-person discrimination

### ISSUE THREAD

**Issue by tuky191** [tuky191/haptic-track#34]
## Problem

Current re-acquisition can't reliably distinguish between two people. MobileNetV3 embeddings are trained for general visual similarity, not within-category identity. Two different people in similar clothing get sim ~0.5-0.7, same as the same person from a different angle.

The cascade label gate prevents cross-category leakage (person locked as chair), but within "person" we have weak discrimination. Current signals:
- MobileNetV3 embedding: not designed for person re-ID
- Color histogram: fails when people wear similar colors
- Person attributes (Crossroad-0230): coarse — gender, clothing type, accessories. Two similar men are indistinguishable.

## Proposed solution: progressive identity building

Build identity signals incrementally during tracking, use the strongest available for re-acquisition:

### Signal hierarchy (strongest → weakest)

1. **Face embedding** (MobileFaceNet or similar, ~1-5MB) — near-certain identity when face visible. Added to lock attributes when face first appears during confirmed tracking.
2. **Body re-ID embedding** (OSNet-AIN or similar, ~2MB) — full-body identity encoding that captures body type, proportions, gait, clothing. Available from first lock.
3. **Person attributes** (Crossroad-0230, existing) — gender, clothing, accessories. Coarse but always available.
4. **Body proportions** (MediaPipe Pose landmarks or bounding box ratios) — shoulder/hip ratio, height/width as proxy for body type (fit/normal/obese).
5. **Clothing color** (existing histogram) — shirt/pants color.
6. **Generic embedding** (MobileNetV3, existing) — general visual similarity.

### Lock lifecycle

```
Initial lock (any angle):
  → body re-ID embedding (OSNet)
  → person attributes (Crossroad-0230)  
  → clothing colors (histogram)
  → generic embedding (MobileNetV3)

During confirmed tracking, face appears:
  → face embedding computed and stored
  → strongest identity signal now available

Re-acquisition:
  → cascade through signals by strength
  → face visible + face embedding stored → high-confidence match
  → no face → fall back to body re-ID + attributes + color
  → nothing matches → clean loss (don't hop to another person)
```

### Key principle

**Clean cutover on loss.** If the best available signals don't match above threshold, accept the loss rather than locking onto a different person. Better to lose tracking than to track the wrong person.

## Research needed

1. **Face embedding models** — MobileFaceNet, ArcFace-Lite, or similar. Need: TFLite compatible, <5MB, works on crops from BlazeFace output (we already detect faces).
2. **Body re-ID models** — OSNet-AIN, TransReID-Light, or similar. Need: TFLite compatible, <5MB, works on person crops.
3. **Body type from pose** — Can MediaPipe Pose landmarks give useful body proportion metrics? Or is a dedicated model better?
4. **Integration with cascade scoring** — How do face/body re-ID embeddings interact with the existing label gate + ranking? Probably: face embedding replaces generic embedding as primary ranking signal for persons.

## Constraints

- Total model size budget: already at ~38MB. Adding 5-10MB for face + body re-ID is acceptable.
- Must work on-device, real-time (or near — can run on lock + every Nth frame like YOLOv8)
- BlazeFace already runs for face detection — face embedding adds one more inference on the crop
- Person attributes pipeline already crops and classifies — body re-ID can share the crop

**Comment by tuky191** [comment]
## Research Results

### Face Embedding: MobileFaceNet (recommended)

| | |
|---|---|
| Size | ~1 MB (TFLite) |
| Input | 112x112 aligned face crop |
| Embedding | 128-dim, L2-normalized |
| Accuracy | 99.55% LFW |
| Inference | ~5 ms on mobile |
| Same person | sim 0.5-0.8 |
| Different people | sim 0.0-0.3 |
| TFLite | Available — `sirius-ai/MobileFaceNet_TF`, `neuralet`, `patlevin/face-search-tflite` |

Alternatives considered: ArcFace-MobileNetV2 (~3 MB, 512-dim, marginal accuracy gain), FaceNet-MobileNetV1 (~3.5 MB). MobileFaceNet is smallest and sufficient.

**Important**: face alignment using BlazeFace eye landmarks (affine warp to canonical position) significantly improves accuracy (~97% → 99%+). Most MobileFaceNet models expect an aligned 112x112 face.

### Body Re-ID: OSNet x1.0 INT8 (recommended)

| | |
|---|---|
| Size | ~2.2 MB (INT8 quantized) |
| Input | 256x128 person crop |
| Embedding | 512-dim |
| Accuracy | 84.9% mAP, 94.8% rank-1 on Market-1501 |
| Inference | ~3-8 ms estimated |
| Same person | sim 0.6-0.8+ across viewpoints |
| Different people | sim 0.1-0.4 |
| TFLite | Needs conversion: PyTorch → ONNX → TFLite (torchreid has ONNX export built in) |

Alternatives considered: OSNet x0.5 (~0.7 MB, lighter but lower accuracy), OSNet x0.25 (~0.25 MB, too weak), TransReID (~27 MB, too large), DeepSORT CNN (~5 MB, weak accuracy).

OSNet-AIN variant preferred for cross-domain generalization (varying lighting/camera conditions).

### MediaPipe Pose for body type: not recommended

Can give coarse body type via shoulder-to-hip ratio from landmarks, but:
- Viewpoint-sensitive (shoulder width appears narrower from side)
- Waist-up crops miss hip landmarks
- OSNet already encodes body shape implicitly in its embedding

Not worth adding when OSNet captures body proportions as part of the re-ID embedding.

### Model budget impact

| Addition | Size |
|---|---|
| MobileFaceNet | +1 MB |
| OSNet x1.0 (INT8) | +2.2 MB |
| **New total** | **~41 MB (from ~38 MB)** |

### Integration plan

**Phase 1: OSNet body re-ID**
1. Export OSNet x1.0 from torchreid → ONNX → TFLite (INT8)
2. New `PersonReIdEmbedder` class, similar to `AppearanceEmbedder`
3. At lock time on "person": compute OSNet embedding from person crop
4. During re-acquisition: use OSNet similarity as primary ranking signal for person candidates (replaces MobileNetV3 for persons)

**Phase 2: MobileFaceNet face identity**
1. Find/convert MobileFaceNet TFLite
2. Add face alignment using BlazeFace eye landmarks (affine warp to 112x112)
3. At lock time: if face visible, compute face embedding
4. During tracking: if face appears, add face embedding to lock (progressive)
5. During re-acquisition: face embedding is highest-priority signal when available

**Phase 3: Cascade integration**
Update `scoreCandidate()` for person candidates:
```
Gate B (label): person label matches → pass
Ranking for persons:
  1. Face embedding (if stored + face visible) — 60%
  2. OSNet body re-ID embedding — 30%  
  3. Person attributes + color — 10%
  
  No face stored or visible:
  1. OSNet body re-ID embedding — 50%
  2. Person attributes + color — 30%
  3. Generic MobileNetV3 — 20%
```

Clean loss: if best person candidate scores below threshold → lose tracking, don't hop to another person.

### Conversion notes (from YOLOv8 experience)

- torchreid provides ONNX export: `torchreid.utils.FeatureExtractor` → `torch.onnx.export()`
- ONNX → TFLite via `onnx2tf` (same tool used for YOLOv8)
- INT8 quantization needs representative dataset (person crops from test footage)
- May need a separate Python venv (torchreid needs specific torch version), similar to `/private/tmp/yolo_export_venv/`

**Comment by tuky191** [comment]
Closed: PR #36 merged — face embedding (MobileFaceNet) + body re-ID (OSNet x1.0) implemented.
